File: src/main/java/me/kgustave/nightfury/commands/BanCmd.java
/*
 * Copyright 2017 Kaidan Gustave
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.kgustave.nightfury.commands;

import java.util.List;
import java.util.stream.Collectors;
import com.jagrosh.jdautilities.commandclient.CommandEvent;

import me.kgustave.nightfury.manager.LogAction;
import me.kgustave.nightfury.manager.NightFuryManager;
import me.kgustave.nightfury.manager.Settings;
import me.kgustave.nightfury.utils.FormatUtils;
import me.kgustave.nightfury.utils.LogUtils;
import net.dv8tion.jda.core.Permission;
import net.dv8tion.jda.core.entities.Member;
import net.dv8tion.jda.core.entities.TextChannel;
import net.dv8tion.jda.core.entities.User;

/**
 *
 * @author Kaidan Gustave
 */
public class BanCmd extends NightFuryCommand {
    
    public BanCmd(NightFuryManager manager) {
        super(manager);
        this.name = "ban";
        this.arguments = "[@user or ID] <reason>";
        this.help = "bans a user from the server";
        this.botPermissions = new Permission[]{Permission.BAN_MEMBERS};
        this.category = manager.MODERATOR;
        this.guildOnly = true;
    }

    @Override
    protected void execute(CommandEvent event) {
        Settings settings = manager.getSettingsForGuild(event.getGuild());
        if(event.getArgs().matches(FormatUtils.TARGET_ID_REASON_REGEX)) {
            String id = event.getArgs().replaceAll(FormatUtils.TARGET_ID_REASON_REGEX, "$2").trim();
            String reason = event.getArgs().replaceAll(FormatUtils.TARGET_ID_REASON_REGEX, "$5").trim();
            event.getJDA().retrieveUserById(id).queue(target -> {
                executeBan(event,settings,target,reason);
            }, err -> {
                event.replyError("User could not be found!");
            });
        } else if(event.getArgs().matches(FormatUtils.TARGET_ID_REASON_FAIL_REGEX)) {
            String name = event.getArgs().replaceAll(FormatUtils.TARGET_ID_REASON_FAIL_REGEX, "$2").trim();
            String descriminator = event.getArgs().replaceAll(FormatUtils.TARGET_ID_REASON_FAIL_REGEX, "$3").trim();
            String reason = event.getArgs().replaceAll(FormatUtils.TARGET_ID_REASON_FAIL_REGEX, "$6").trim();
            List<Member> members = event.getGuild().getMembersByName(name, false)
                    .stream().filter(member -> member.getUser().getDiscriminator().equals(descriminator)).collect(Collectors.toList());
            if(members.size()<1)
                event.replyError("No users detected going by **"+name+"**#"+descriminator+"!");
            else if(members.size()>1)
                event.replyError("Multiple users detected going by **"+name+"**#"+descriminator+"!");
            else
                executeBan(event,settings,members.get(0).getUser(),reason);
        } else {
            event.replyError(FormatUtils.invalidArgs(this.getName()));
        }
    }
    
    private void executeBan(CommandEvent event, Settings settings, User target, String reason) {
        String user = FormatUtils.formattedUserName(target, true);
        if(event.getSelfUser().equals(target))
            event.replyError("I cannot ban myself from the server!");
        else if(event.getAuthor().equals(target))
            event.replyError("I cannot ban you from the server!");
        else if(event.getGuild().getMember(target)!=null && !event.getSelfMember().canInteract(event.getGuild().getMember(target)))
            event.replyError("I cannot ban "+user+" because they have a higher role than me!");
        else {
            event.getGuild().getController().ban(target, 1).reason((reason.isEmpty()? "No reason specified" : reason)).queue(v -> {
                event.replySuccess(user+" was banned from the server.");
                if(settings.hasModLogId()) {
                    TextChannel log = event.getGuild().getTextChannelById(settings.getModLogId());
                    if(!event.getSelfMember().hasPermission(log, Permission.MESSAGE_WRITE))
                        return;
                    else if(!reason.isEmpty())
                        LogUtils.userWithReason(settings, log, event.getAuthor(), target, LogAction.BAN, reason);
                    else
                        LogUtils.userWithNoReason(settings, log, event.getAuthor(), target, LogAction.BAN);
                }
            }, err -> {
                event.replyError("An unexpected error occurred while banning "+user+"!");
            });
        }
    }
}

File: src/main/kotlin/xyz/nightfury/commands/moderator/BanCmd.kt
/*
 * Copyright 2017-2018 Kaidan Gustave
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package xyz.nightfury.commands.moderator

import xyz.nightfury.Category
import xyz.nightfury.Command
import xyz.nightfury.CommandEvent
import xyz.nightfury.annotations.MustHaveArguments
import xyz.nightfury.entities.promise
import xyz.nightfury.extensions.banFrom
import xyz.nightfury.extensions.formattedName
import net.dv8tion.jda.core.Permission
import xyz.nightfury.annotations.HasDocumentation
import xyz.nightfury.db.entities.ModLogger

/**
 * @author Kaidan Gustave
 */
@HasDocumentation
@MustHaveArguments("Mention a user or provide a user ID to ban.")
class BanCmd : Command() {
    init {
        this.name = "Ban"
        this.arguments = "[@User or ID] <Reason>"
        this.help = "Bans a user from the server."
        this.botPermissions = arrayOf(Permission.BAN_MEMBERS)
        this.category = Category.MODERATOR
        this.guildOnly = true
    }

    override fun execute(event: CommandEvent) {
        val parsed = event.modSearch() ?: return

        val id = parsed.first
        val reason = parsed.second

        event.jda.retrieveUserById(id).promise() then { user ->
            // If no user is found we just respond and end.
            if(user == null)
                return@then event.replyError("Could not find a user matching \"${event.args}\"!")

            // Error Responses
            val error = when {
                event.selfUser == user
                        -> "I cannot ban myself from the server!"

                event.author == user
                        -> "You cannot ban yourself from the server!"

                event.guild.owner.user == user
                        -> "You cannot ban ${user.formattedName(true)} because they are the owner of the server!"

                event.guild.isMember(user) && !event.selfMember.canInteract(event.guild.getMember(user))
                        -> "I cannot ban ${user.formattedName(true)}!"

                event.guild.isMember(user) && !event.member.canInteract(event.guild.getMember(user))
                        -> "You cannot ban ${user.formattedName(true)}!"

                else    -> null
            }

            if(error != null)
                return@then event.replyError(error)

            user.banFrom(event.guild, 1, reason).promise() then {
                ModLogger.newBan(event.member, user, reason)
                event.replySuccess("${user.formattedName(true)} was banned from the server.")
            } catch {
                event.replyError("Banning ${user.formattedName(true)} failed for an unexpected reason!")
            }
        } catch {
            event.replyError("An unexpected error occurred when finding a user with ID: $id!")
        }
    }
}
